### PR TITLE
Implementation of the logistic-normal distribution

### PR DESCRIPTION
**Pull request by alshedivat**
Implements the [logistic-normal distribution](https://en.wikipedia.org/wiki/Logit-normal_distribution).

**Note:** the moments of the distribution don't have an analytical form, hence the class will throw `NotImplementedError` when `mean` or `variance` is called.  Perhaps it would be better to throw a more specific exception (please suggest).

### CODE REVIEW

**Comment by vishwakftw** [comment]
cc @fritzo @alicanb

File: torch/distributions/transforms.py [hunk 1]
```diff
@@ -412,6 +412,29 @@ def _inverse(self, y):
         return probs.log()
 
 
+class InvertableBoltzmannTransform(Transform):
+    """
+    Transform from unconstrained space to the simplex via `y = [exp(x), 1]` and
+    then normalizing. Note that dimensionality of `y.dim() == x.dim() + 1`.
```
Comment (fritzo) [inline torch/distributions/transforms.py]: I think the dimensionality change will violate an assumption in `test_distributions.py`, so you'll need special handling when adding `InvertableBoltzmannTransform` to the tests.

**Comment by vishwakftw** [comment]
If I understand correctly, this can be implemented using existing transforms. According to the Wiki page:

> `X = P(Y)` where Y ~ Normal and P ~ Sigmoid

**Review by fritzo — CHANGES_REQUESTED** [review]
The `LogitNormal` transform seems pretty simple to implement once we have `InvertibleBoltzmannTransform`.

To Do:
- add tests for the transform
- add tests for the distribution
- add docs for the distribution

**Comment by fritzo** [comment]
@alshedivat to see how to add tests, search tests/test_distributions.py for `LogNormal` and `StickBreakingTransform` and add entries for `LogitNormal` and `InvertibleBoltzmannTransform`. To see how to add docs, search docs/source/distributions.rst for `LogNormal` (I believe the transform will be registered in docs automatically).

File: torch/distributions/transforms.py [hunk 2]
```diff
@@ -412,6 +412,29 @@ def _inverse(self, y):
...
+
+    def _call(self, x):
+        ones = torch.ones(*(x.size()[:-1] + (1, )))
+        probs = torch.cat([x.exp(), ones], dim=-1)
+        probs /= probs.sum(-1, True)
+        return probs
+
+    def _inverse(self, y):
+        logprobs = y[:, :-1].log() - y[:, -1:].log()
+        return logprobs
```
Comment (apaszke) [inline torch/distributions/transforms.py]: can you compress this to a single `return`? the assignment is unnecessary

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -412,6 +412,29 @@ def _inverse(self, y):
...
+    """
+    domain = constraints.real
+    codomain = constraints.simplex
+    event_dim = 1
+
+    def __eq__(self, other):
+        return isinstance(other, InvertableBoltzmannTransform)
+
+    def _call(self, x):
+        ones = torch.ones(*(x.size()[:-1] + (1, )))
```
Comment (apaszke) [inline torch/distributions/transforms.py]: This won't be on the correct device. Not sure what's the best one liner we could use right now, I'll need to think

File: torch/distributions/logit_normal.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        super(LogitNormal, self).__init__(
+            Normal(loc, scale), InvertableBoltzmannTransform())
+
+    @property
+    def loc(self):
+        return self.base_dist.loc
+
+    @property
+    def scale(self):
+        return self.base_dist.scale
```
Comment (apaszke) [inline torch/distributions/logit_normal.py]: @fritzo Given how simple this is to implement using the transform, do we want to merge it, or just keep the transform?

File: torch/distributions/transforms.py (hunk 1 above)
Comment (alicanb) [reply torch/distributions/transforms.py]: nit: Can we call it `InvertibleBoltzmannTransform` instead of `InvertableBoltzmannTransform`?

**Comment by alshedivat** [comment]
@vishwakftw Re: using existing transforms -- it is possible for the univariate logit-normal distribution, but not for the multivariate generalization. Basically, using `X = P(Y)` where `Y ~ Normal` and `P ~ Sigmoid` makes each component of `X` be an element of a 2-dim simplex. `InvertibleBoltzmannTransform` maps a `(D-1)`-dim vector onto a `D`-dim simplex (i.e., elements of `X` sum to 1).

To avoid confusion, I'm thinking to rename the distribution to `LogisticNormal` (also consistent with the [wiki page](https://en.wikipedia.org/wiki/Logit-normal_distribution#Multivariate_generalization)) and perhaps add `LogitNormal` as a separate one (implemented using the `Sigmoid` transform). Although, as @apaszke mentioned, is it worth having separate distributions or better add only transforms? What are the current guidelines for this?

File: torch/distributions/transforms.py (hunk 1 above)
Comment (alshedivat) [reply torch/distributions/transforms.py]: `InvertableBoltzmannTransform` -> `InvertibleBoltzmannTransform ` done.

File: torch/distributions/transforms.py (hunk 2 above)
Comment (alshedivat) [reply torch/distributions/transforms.py]: Done.

File: torch/distributions/transforms.py (hunk 3 above)
Comment (alshedivat) [reply torch/distributions/transforms.py]: How about `probs = F.pad(x.exp(), (0, 1), value=1.)` instead of lines 428-29?

File: torch/distributions/transforms.py [hunk 5]
```diff
@@ -412,6 +413,31 @@ def _inverse(self, y):
...
+    domain = constraints.real
+    codomain = constraints.simplex
+    bijective = True
+    event_dim = 1
+
+    def __eq__(self, other):
+        return isinstance(other, InvertibleBoltzmannTransform)
+
+    def _call(self, x):
+        probs = pad(x.exp(), (0, 1), value=1)
```
Comment (apaszke) [inline torch/distributions/transforms.py]: Oh, that's perfect!

**Comment by ezyang** [comment]
Do we have a testing strategy for distributions?

**Comment by vishwakftw** [comment]
Tests have not been added in this PR. @alshedivat . Generally, there is a scipy-based testing framework used. For entropy and KL-divergences, a Monte-Carlo approximation is used.

**Comment by alshedivat** [comment]
@vishwakftw, sounds good. Yes, I looked into how testing of transformed distributions is done. Will add tests in the next couple of days.

**Comment by alshedivat** [comment]
Updates:
- `LogitNormal` and `LogisticNormal` are different distributions now (see comments above).
- Removed `InvertibleBoltzmannTransform` and implemented `LogisticNormal` using the `StickBreakingTransform` (this is cleaner and consistent with [Stan](http://mc-stan.org/)).
- Implemented `log_abs_det_jacobian`for the `StickBreakingTransform` (necessary for computing `LogisticNormal.log_prob`).
- Added tests.

File: torch/distributions/kl.py [hunk 6]
```diff
@@ -313,6 +315,13 @@ def _kl_transformed_transformed(p, q):
     return kl_divergence(p.base_dist, q.base_dist)
 
 
+@register_kl(LogisticNormal, LogisticNormal)
```
Comment (fritzo) [inline torch/distributions/kl.py]: This should not be needed. `kl_divertence(LogisticNormal(...), LogisticNormal(...))` should come for free because it inherits from `TransformedDistribution`. Could you remove this registration?

File: torch/distributions/logistic_normal.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    def __init__(self, loc, scale):
+        super(LogisticNormal, self).__init__(
+            Normal(loc, scale), StickBreakingTransform())
+        # Adjust event shape since StickBreakingTransform adds 1 dimension
+        self._event_shape = self._event_shape.__class__([
+            s + 1 for s in self._event_shape
+        ])
+
+    def log_prob(self, value):
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: I'm inclined to fix `TransformedDistribution.log_prob()` rather than duplicate most of its logic here. Does the following fix work for you?
```diff
  class TransformedDistribution(Distribution):
      ...
      def log_prob(self, value):
          """
          Scores the sample by inverting the transform(s) and computing the score
          using the score of the base distribution and the log abs det jacobian.
          """
-         self.base_dist._validate_log_prob_arg(value)
+         self._validate_log_prob_arg(value)
          event_dim = len(self.event_shape)
          log_prob = 0.0
          y = value
          for transform in reversed(self.transforms):
              x = transform.inv(y)
              log_prob -= _sum_rightmost(transform.log_abs_det_jacobian(x, y),
                                         event_dim - transform.event_dim)
              y = x
          log_prob += _sum_rightmost(self.base_dist.log_prob(y),
                                     event_dim - len(self.base_dist.event_shape))
          return log_prob
```
If so then you should be able to delete this method here.

File: torch/distributions/logistic_normal.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    params = {'loc': constraints.real, 'scale': constraints.positive}
+    support = constraints.simplex
+    has_rsample = True
+
+    def __init__(self, loc, scale):
+        super(LogisticNormal, self).__init__(
+            Normal(loc, scale), StickBreakingTransform())
+        # Adjust event shape since StickBreakingTransform adds 1 dimension
+        self._event_shape = self._event_shape.__class__([
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: I would replace `self._event_shape.__class__` with `torch.Size`.

File: torch/distributions/logit_normal.py (hunk 4 above)
Comment (fritzo) [reply torch/distributions/logit_normal.py]: I'm happy either way, but I guess I don't see much value over the one-line version:
```py
TransformedDistribution(Normal(loc, scale), SigmoidTransform())
```
@alshedivat do you have specific need for the extra constraints and attributes provided by subclassing? Note that `kl_divergence(-,-)` should already come for free.

File: torch/distributions/transforms.py [hunk 9]
```diff
@@ -434,21 +434,24 @@ def __eq__(self, other):
...
-        units = y[..., :-1] / sf[..., :-1]
-        return units.log()
+        shape = y.shape[:-1] + (y.shape[-1] - 1,)
+        offset = (shape[-1] + 1) - y.new([1]).expand(shape).cumsum(-1)
+        sf = (1 - y.cumsum(-1))[..., :-1]
+        x = y[..., :-1].log() - sf.log() + offset.log()
+        return x
 
-    # TODO implement .log_abs_det_jacobian()
+    def log_abs_det_jacobian(self, x, y):
```
Comment (fritzo) [inline torch/distributions/transforms.py]: 🎉 Thank you for implementing this!

How did you verify its correctness? We don't have great tests for multivariate `log_abs_det_jacobian`.

**Review by fritzo — COMMENTED** [review]
Nice tests!

File: torch/distributions/transforms.py (hunk 9 above)
Comment (alshedivat) [reply torch/distributions/transforms.py]: My implementation is based on [Stan's reference manual](https://github.com/stan-dev/stan/releases/) (Section 35.6, pp. 408-410).

Yes, I noticed that `log_abs_det_jacobian` is tested only for univariate cases. Not sure if there is a way to have a generic set of tests for multivariate `log_abs_det_jacobian`. An indirect way would be to test `log_prob` of a distribution that uses this transform against an existing implementation. Unfortunately, `scipy.stats` doesn't have logistic-normal distribution.

File: torch/distributions/logit_normal.py (hunk 4 above)
Comment (alshedivat) [reply torch/distributions/logit_normal.py]: @fritzo, agreed -- one-line version should work and the extra constraints should come from the transform for free. Given that moments for this distribution don't have an analytical form, it's probably not worth subclassing.

File: torch/distributions/logistic_normal.py (hunk 7 above)
Comment (alshedivat) [reply torch/distributions/logistic_normal.py]: This is perfect!

File: torch/distributions/kl.py [hunk 10]
```diff
@@ -310,7 +313,11 @@ def _kl_poisson_poisson(p, q):
 def _kl_transformed_transformed(p, q):
     if p.transforms != q.transforms:
         raise NotImplementedError
-    return kl_divergence(p.base_dist, q.base_dist)
+    if p.event_shape != q.event_shape:
+        raise NotImplementedError
+    event_dim = len(p.event_shape)
```
Comment (fritzo) [inline torch/distributions/kl.py]: I think you'll need to account for cases where `p.base_dist` is itself multivariate:
```py
extra_event_dim = len(p.event_shape) - len(p.base_dist.event_shape)
...
_sum_rightmost(base_kl_divergence, extra_event_dim)
```

File: torch/distributions/logistic_normal.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.normal import Normal
+from torch.distributions.transformed_distribution import TransformedDistribution
+from torch.distributions.transforms import ComposeTransform, ExpTransform, StickBreakingTransform
+from torch.distributions.utils import _sum_rightmost
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: I think flake8 wants you to delete this line

File: torch/distributions/logit_normal.py (hunk 4 above)
Comment (fritzo) [reply torch/distributions/logit_normal.py]: Correct, you can delete this file and its tests. Thanks for writing the tests though!

**Comment by alshedivat** [comment]
FYI (unrelated to this PR), [`test_cdf_icdf_inverse`](https://github.com/pytorch/pytorch/blob/master/test/test_distributions.py#L1312-L1329) fails for one of the `Pareto` distributions in `EXAMPLES`. This doesn't happen always but when `actual` or `samples` contain `inf` which results into `nan` in `rel_error` and fails `assertLess`.

**Comment by vishwakftw** [comment]
@alshedivat If `samples` contained `inf`, then `rel_error` would be `0.0`. However, this problem can exist if `actual` was `inf` too, since in Python `inf / inf == nan`. @fritzo any comments?

**Comment by fritzo** [comment]
> `test_cdf_icdf_inverse` fails

This just sounds like numerical imprecision. My only concern is that it sounds like rng seeds aren't being set deterministically.

**Comment by alshedivat** [comment]
Rebased this PR against the most recent `master`. The issue with NaNs still persist in `test_cdf_icdf_inverse`, but I think that should be solved by a separate PR.

**Comment by apaszke** [comment]
@alshedivat it looks like you made some mistakes when rebasing your commits (many `master` commits show up as part of this PR). Can you please try again?

**Comment by alshedivat** [comment]
@apaszke, for some reason, rebase kept adding commit history from `master`. I ended up using the standard merge instead.

File: torch/distributions/logistic_normal.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class LogisticNormal(TransformedDistribution):
+    r"""
+    Creates a logistic-normal distribution parameterized by
+    `mean` and `std` where::
+
+        X ~ LogisticNormal(loc, scale)
+        Y = log(X / (1 - X)) ~ Normal(loc, scale)
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: This math confuses me; it suggests that `X` is univariate, but you've implemented the more general multivariate version (e.g Bernoulli vs Categorical). Is there a simple way to define your `LogisticNormal` distribution?

File: torch/distributions/logistic_normal.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    Example::
+
+        >>> m = LogisticNormal(torch.Tensor([0.0]), torch.Tensor([1.0]))
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: Could you add a higher dimensional example, say with 3 inputs and 4 outputs? It's not clear to me from this documentation what the `loc` and `scale` parameters mean.

File: torch/distributions/logistic_normal.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+from torch.distributions.normal import Normal
+from torch.distributions.transformed_distribution import TransformedDistribution
+from torch.distributions.transforms import ComposeTransform, ExpTransform, StickBreakingTransform
+from torch.distributions.utils import _sum_rightmost
+
+
+class LogisticNormal(TransformedDistribution):
+    r"""
+    Creates a logistic-normal distribution parameterized by
+    `mean` and `std` where::
```
Comment (fritzo) [inline torch/distributions/logistic_normal.py]: Could you replace `mean` and `std` with `loc` and `scale`? Feel free to fix the stale docstrings from which you copied these too :wink:

**Review by fritzo — APPROVED** [review]
LGTM.  @apaszke could you PTAL?

File: test/test_distributions.py [hunk 15]
```diff
@@ -2968,7 +3052,9 @@ def test_univariate_forward_jacobian(self):
 
     def test_univariate_inverse_jacobian(self):
         for transform in self.transforms:
-            y = variable(self._generate_data(transform.inv), requires_grad=True)
+            if transform.event_dim > 0:
+                continue
+            y = Variable(self._generate_data(transform.inv), requires_grad=True)
```
Comment (apaszke) [inline test/test_distributions.py]: Please revert this to use `variable(...)` (same above)

**Review by apaszke — COMMENTED** [review]
Minor fix needed. Looks good

**Comment by apaszke** [comment]
Also looks like it's conflicting with some other changes in `master` now

**Comment by apaszke** [comment]
It still conflicts with master (it's possible that it's because of the PR I merged a moment ago, sorry!) 😕

**Comment by alshedivat** [comment]
@apaszke, resolved.

**Comment by apaszke** [comment]
@alshedivat looks like the tests are failing

**Comment by alshedivat** [comment]
@apaszke, this is the problem in `Pareto` distribution that I mentioned above (see [logs](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py3.6-gcc5.4-test/3468/console) ), which is unrelated to this PR.

**Comment by alshedivat** [comment]
@vishwakftw, nice, it worked out on my machine.

**Comment by vishwakftw** [comment]
@fritzo I think it is better to adopt a streaming approach as done for the KL divergence tests. What do you think?

**Review by fritzo — APPROVED** [review]

**Comment by fritzo** [comment]
> a streaming approach as done for the KL divergence tests

That's reasonable if the `.entropy()` tests are getting expensive. Let's not hold up @alshedivat's PR though :smile:

File: torch/distributions/kl.py [hunk 16]
```diff
@@ -310,7 +311,12 @@ def _kl_poisson_poisson(p, q):
 def _kl_transformed_transformed(p, q):
     if p.transforms != q.transforms:
         raise NotImplementedError
-    return kl_divergence(p.base_dist, q.base_dist)
+    if p.event_shape != q.event_shape:
+        raise NotImplementedError
+    # extra_event_dim = len(p.event_shape) - len(p.base_dist.event_shape)
+    extra_event_dim = len(p.event_shape)
```
Comment (fritzo) [inline torch/distributions/kl.py]: oops, looks like the commented-out line is correct, as pointed out in #34859.
